src/resources/uart_esp32.cc

```diff
@@ -888,24 +913,28 @@ PRIMITIVE(wait_tx) {
 
   if (!uart->tx_buffer()->is_empty()) return BOOL(false);
 
-  // Busy wait for the fifo to become empty.
-  while (uart->get_tx_fifo_available_count() < SOC_UART_FIFO_LEN) ;
+  while (uart->get_tx_fifo_available_count() < SOC_UART_FIFO_LEN) {
+    // Busy wait for the fifo to become empty.
+  }
   return BOOL(true);
 }
 
 PRIMITIVE(read) {
   ARGS(UartResource, uart)
 
-  uword available = uart->rx_buffer()->available_size();
+  RxBuffer* buffer = uart->rx_buffer();
+  uword available = buffer->available_size();
+  if (available == 0) return process->program()->null_object();
 
-  ByteArray* data = process->allocate_byte_array(static_cast<int>(available), /*force_external*/ available != 0);
+  // TODO(kasper): It isn't obviously a good idea to just return
+  // all the data in a potentially rather large external byte array.
+  // For reads from TCP sockets, we chunk it up instead and prefer
+  // to return multiple smaller byte arrays.
+  ByteArray* data = process->allocate_byte_array(available);
```

Comment:
We probably need to think about sticking to external byte arrays for larger reads, if that is necessary for performance. @mikkeldamsgaard might know.

Reply:
Maximum buffer size is 2k, so there should be no problem in having this not external.